### PR TITLE
Add Support for CentOS 7+

### PR DESCRIPTION
**Pull request by alexmbird**
I've added support for installing scollector on CentOS 7 and above.  It drops a systemd service definition into place and starts scollector.

Right now it's guarded with a case statement that it'll **only** run for CentOS 7 & above.  In all other cases it falls back to your old runit method.  We may wish to add cases for other distributions that have moved to systemd.

### CODE REVIEW

**Comment by ptqa** [comment]
As I can see you've managed to make kitchen.yml by yourself, that's great.

Can you please squash it all into 1 commit? 
Is kitchen test works without errors after your commits?

**Comment by alexmbird** [comment]
I've split the kitchen tests into separate suites for CentOS and Ubuntu.  Your original default is renamed to ubuntu.  While doing this I've run into a problem: my new CentOS tests pass fine but the ubuntu ones (your original default) have been broken for a while.  I think it may be an upstream problem with the `runit` cookbook.

```
         * runit_service[scollector] action restart

           ================================================================================
           Error executing action `restart` on resource 'runit_service[scollector]'
           ================================================================================

           Errno::ENOENT
           -------------
           No such file or directory - /usr/bin/sv
```

It's possible that my .kitchen.yml is missing something important (runit attributes?) to make the Ubuntu tests work.  Can I see a copy?

**Comment by ptqa** [comment]
Basically you should have kitchen.yml like that https://gist.github.com/ptqa/63c440e5fc6cd645d488

I do not have any problem with /usr/bin/sv, but for some reason now I can't install scollector with go packages when I run "kitchen test", but it works fine when I run "kitchen converge" .... 

Will try to investigate this. 

P.S. Also I think there should be no "service_type" variables, there should be "init_style" attribute. And you can set it to "systemd" in case of centos, but user may override this.

**Comment by alexmbird** [comment]
> Also I think there should be no "service_type" variables, there should be "init_style" attribute. And you can set it to "systemd" in case of centos, but user may override this

If we were setting up regular services with a `service` resource for both distros that would be fine.  But for Ubuntu support you've used a `runit_service` and I need a way to switch between them.

Any ideas?

**Comment by ptqa** [comment]
> But for Ubuntu support you've used a runit_service and I need a way to switch between them.
> It's not a big deal actually. You can do something like:

```
notifies :restart, node['scollector']['init_style'] == 'runit' ? 'runit_service[scollector]' : 'service[scollector]'
```

Or even you can leave your service_resource variable. The point is not to use service_type as variable, but as attribute.

**Comment by ptqa** [comment]
Another option is to subscribe service to template like:

```
runit_service 'scollector' do
  cookbook node['scollector']['config_cookbook']
  restart_on_update true
  subscribes :restart, "template[#{node['scollector']['conf_dir']}/scollector.toml]"
end
```

**Comment by alexmbird** [comment]
Good idea, `subscribes` will simplify it a lot.

But we still need to know whether to create a `service` or a `runit_service`.  I can switch that based upon the value of `default['scollector']['init_style']` but then the user would need to explicitly set that attribute to get the correct setup for their flavour of Linux instead of having it automatically.  I think automatic would be a lot more convenient.

I've altered the code to try and give us the best of both worlds.  If `default['scollector']['init_style']` has been set by the user it is honoured, otherwise it uses my logic to infer the system automatically.  Does that sound okay to you?

**Comment by ptqa** [comment]
Yeah, something like that. Can you please check that service really restart when you change template?

Also I think it would be better (and more simple) to put your logic of choosing init style to attributes/default.rb, like this:

```
if node['platform'] === 'centos' && platform_major_num >= 7
  default['scollector']['init_style'] = 'systemd'
else
  default['scollector']['init_style'] = 'runit'
end
```

That should work, but I did not tested it.

**Comment by alexmbird** [comment]
Okay, I've moved the logic into default.rb.

Do you know of a way to test whether a service has been restarted?  Nothing obvious in the [serverspec docs](http://serverspec.org/resource_types.html#service).

**Comment by ptqa** [comment]
This look good to me.

I don't think you can test it with server spec. Ok, I'll try to do it by myself.

**Comment by alexmbird** [comment]
Thanks.

Tests for the runit variant are still failing I'm afraid:

```
$ kitchen test ubuntu-14

...

================================================================================
Error executing action `create` on resource 'link[/etc/service/scollector]'
================================================================================

Errno::ENOENT
-------------
No such file or directory @ sys_fail2 - (/etc/sv/scollector, /etc/service/scollector)

Resource Declaration:
---------------------
# In /tmp/kitchen/cookbooks/runit/libraries/provider_runit_service.rb

241:         link "#{service_dir_name}" do
242:           to sv_dir_name
243:           action :create
244:         end
245:
```

File: recipes/configure.rb [hunk 1]
```diff
@@ -14,7 +14,22 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-include_recipe 'runit'
+
```
Comment (ptqa) [inline recipes/configure.rb:17]: Well, you won't get runit installed if you remove 'include_recipe runit" ;) That's the reason of /usr/bin/sv not found

Should be changed to something like

```
include_recipe 'runit' if node['scollector']['init_style'] == 'runit'
```

**Comment by alexmbird** [comment]
> Well, you won't get runit installed if you remove 'include_recipe runit" ;)

D'oh!  Well spotted.

I've fixed it and now we have an error from runit...

```
================================================================================
     Error executing action `start` on resource 'runit_service[scollector]'
     ================================================================================

 Mixlib::ShellOut::ShellCommandFailed
     ------------------------------------
     Expected process to exit with [0], but received '1'
     ---- Begin output of /usr/bin/sv start /etc/service/scollector ----
     STDOUT: warning: /etc/service/scollector: unable to open supervise/ok: file does not exist
     STDERR:

     Ran /usr/bin/sv start /etc/service/scollector returned 1

     Cookbook Trace:
     ---------------
     /tmp/kitchen/cookbooks/runit/libraries/helpers.rb:166:in `start_service'
     /tmp/kitchen/cookbooks/runit/libraries/provider_runit_service.rb:281:in `block in <class:RunitService>'

     Resource Declaration:
 ---------------------
     # In /tmp/kitchen/cookbooks/scollector/recipes/configure.rb

      81:   runit_service 'scollector' do
      82:     cookbook    node['scollector']['config_cookbook']
      83:     restart_on_update true
      84:     subscribes  :restart, "template[#{node['scollector']['conf_dir']}/scollector.toml]"
      85:     action      [:enable, :start]
      86:   end
      87:
```

**Comment by ptqa** [comment]
@alexmbird actually I'm not sure that we need to start runit service manually. Isn't :enable enough here? AFAIK runsv-dir should start process after :enable.

**Comment by alexmbird** [comment]
Interesting, I'l give that a go

File: recipes/configure.rb [hunk 2]
```diff
@@ -33,10 +48,43 @@
...
+  runit_service 'scollector' do
+    cookbook    node['scollector']['config_cookbook']
+    restart_on_update true
+    subscribes  :restart, "template[#{node['scollector']['conf_dir']}/scollector.toml]"
+    action      [:enable, :start]
+  end
+
+else
+  log "Unsupported platform"
```
Comment (ptqa) [inline recipes/configure.rb:89]: running rubocop on this file would be nice

**Comment by alexmbird** [comment]
You'll have to forgive me here - I'm not a Ruby developer by trade, just a DevOps guy trying to get scollector onto his CentOS hosts.  Are there any specific things you need fixing?

**Comment by ptqa** [comment]
@alexmbird no problem then, it's just some ruby style waning on rubocop, I'll deal with that after merging.

**Comment by alexmbird** [comment]
With the `:enable` only I'm getting:

```
* runit_service[scollector] action restart

  ================================================================================
  Error executing action `restart` on resource 'runit_service[scollector]'
  ================================================================================

  Mixlib::ShellOut::ShellCommandFailed
  ------------------------------------
  Expected process to exit with [0], but received '1'
  ---- Begin output of /usr/bin/sv restart /etc/service/scollector ----
  STDOUT: warning: /etc/service/scollector: unable to open supervise/ok: file does not exist
  STDERR:
  ---- End output of /usr/bin/sv restart /etc/service/scollector ----
  Ran /usr/bin/sv restart /etc/service/scollector returned 1

  Cookbook Trace:
  ---------------
  /tmp/kitchen/cookbooks/runit/libraries/helpers.rb:174:in `restart_service'
  /tmp/kitchen/cookbooks/runit/libraries/provider_runit_service.rb:277:in `block in <class:RunitService>'

  Resource Declaration:
  ---------------------
  # In /tmp/kitchen/cookbooks/scollector/recipes/configure.rb

   81:   runit_service 'scollector' do
   82:     cookbook    node['scollector']['config_cookbook']
   83:     restart_on_update true
   84:     subscribes  :restart, "template[#{node['scollector']['conf_dir']}/scollector.toml]"
   85:     action      :enable
   86:   end
   87:
```

So I agree the runit cookbook is taking that `:enable` to imply `:start` but it's still breaking in the same way.

**Comment by ptqa** [comment]
Looks like they've broken something in runit cookbook, if you put in your Berksfile:

```
source "https://supermarket.getchef.com"
metadata
cookbook 'apt'
cookbook 'runit', '=1.6.0'
cookbook 'golang', git: 'git@github.com:NOX73/chef-golang.git'
```

There is no error. Or maybe there were some breaking changes and I should not use cookbook this way...

**Comment by alexmbird** [comment]
Okay - let's consider that outside the scope of my work.

Give me a few minutes to tidy up the pull request and I'll be ready to merge.

**Comment by alexmbird** [comment]
Done!  It's all yours

**Comment by ptqa** [comment]
Had time to make little fixes only today. @alexmbird thanks for contribution.

**Pull request merged**